docs/architecture/cloud-native/rest-grpc.md

```diff
@@ -1,69 +1,347 @@
 
-The Microsoft .NET Core framework 3.0 includes tooling and native support for gRPC. Figure 4-20 shows the Visual Studio 2019 template that scaffolds a gRPC skeleton project for a gRPC service. Note how .NET Core supports the Windows, Linux, and macOS platforms.
+Microsoft .NET Core 3.0 includes tooling and native support for gRPC. gRPC has been seamlessly integrated into the 3.0 framework with support for endpoint routing, built-in IoC, and logging. The open-source Kestrel web server supports HTTP/2 connections. Figure 4-20 shows a Visual Studio 2019 template that scaffolds a skeleton project for a gRPC service. Note how .NET Core fully supports the Windows, Linux, and macOS platforms.
 
 ![gRPC Support in Visual Studio 2019](./media/visual-studio-2019-grpc-template.png)
 
 **Figure 4-20**. gRPC support in Visual Studio 2019
-
-.NET Core 3.0 seamlessly integrates gRPC into its framework, including endpoint routing, built-in IoC support, and logging. The open-source Kestrel web server fully supports HTTP/2 connections.
-
-Figure 4-21 shows structure of a gRPC service in Visual Studio 2019. Note how the folder structure includes folders for the proto files and service code.
+  
+Figure 4-21 shows the skeleton gRPC service generated from the built-in scaffolding included in Visual Studio 2019.  
 
 ![gRPC project in Visual Studio 2019](./media/grpc-project.png  )
 
 **Figure 4-21**. gRPC project in Visual Studio 2019
 
-## gRPC Usage
+In the previous figure, note the proto description file and service code. As you'll see shortly, Visual Studio generates additional configuration in both the Startup class and underlying project file.
+
+An excellent introduction to building a gRPC service is [gRPC with ASP.NET Core 3.0](https://www.dotnetcurry.com/aspnet-core/1514/grpc-asp-net-core-3) from the .NET Curry magazine series.
+
+## gRPC usage
+
+Favor gRPC for the following scenarios:
+
+- Synchronous backend microservice-to-microservice communication where an immediate response is required to continue processing.
+- Polyglot environments that need to support mixed programming platforms.
+- Low latency and high throughput communication where performance is critical.
+- Point-to-point real-time communication - gRPC can push messages in real time without polling and has excellent support for bi-directional streaming.
+- Network constrained environments – binary gRPC messages are always smaller than an equivalent text-based JSON message.
+
+At the time of this writing, gRPC is primarily used in backend services. Most modern browsers can't provide the level of HTTP/2 control required to support a front-end gRPC client.
+
+## gRPC implementation
+
+The microservice reference architecture for [eShop on Containers](https://github.com/dotnet-architecture/eShopOnContainers) demonstrates how to implement gRPC into microservices on the .NET Core platform. Figure 4-22 shows the implementation.
+
+![gRPC in eShop on Containers](./media/grpc-implementation.png)
```

Comment:
Use a better diagram here. The important aspect here is to show how aggregators use gRPC.

Reply:
You can briefly mention the service mesh implementation too. Those are practical scenarios.

Reply:
Check this : https://github.com/dotnet-architecture/eShopOnContainers/wiki/API-gateways

Reply:
Refactored.